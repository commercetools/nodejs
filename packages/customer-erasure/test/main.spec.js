import silentLogger from '../src/utils/silent-logger'
import CustomerErasure from '../src/main'

describe('CustomerErasure', () => {
  const logger = {
    ...silentLogger,
  }

  let customerErasure
  beforeEach(async () => {
    customerErasure = new CustomerErasure(
      {
        apiConfig: {
          projectKey: 'test-project-key',
        },
      },
      logger
    )
  })

  describe('::constructor', () => {
    test('should be a function', () => {
      expect(typeof CustomerErasure).toBe('function')
    })

    test('should set default properties', () => {
      expect(customerErasure.apiConfig).toEqual({
        projectKey: 'test-project-key',
      })
      expect(customerErasure.logger).toEqual(logger)
    })

    test('should throw error if no `apiConfig` in `options` parameter', () => {
      expect(
        () => new CustomerErasure({ foo: 'bar' })
      ).toThrowErrorMatchingSnapshot()
    })
  })

  describe('::getCustomerData', () => {
    let payload
    describe('with status code 200', () => {
      beforeEach(() => {
        payload = {
          statusCode: 200,
          body: {
            results: [{ version: 1, id: 'id1' }, { version: 1, id: 'id2' }],
          },
        }
        customerErasure.client.execute = jest.fn(() => Promise.resolve(payload))
      })

      test('should fetch data', async () => {
        const data = await customerErasure.getCustomerData('customerId')
        expect(data).toMatchSnapshot()
      })
    })

    describe('with status code 500', () => {
      beforeEach(() => {
        payload = {
          statusCode: 500,
          body: {
            results: [],
          },
        }
        customerErasure.client.process = jest.fn(async (request, callback) => {
          await callback(payload)
        })
      })

      test('should throw internal server error', async () => {
        expect(
          customerErasure.getCustomerData('customerId')
        ).rejects.toThrowErrorMatchingSnapshot()
      })
    })

    describe('with status code 404', () => {
      beforeEach(() => {
        payload = {
          statusCode: 404,
          body: {
            results: [],
          },
        }
        customerErasure.client.execute = jest.fn(() => Promise.resolve(payload))
      })
      test('should fetch empty data', async () => {
        const data = await customerErasure.getCustomerData('customerId')
        expect(data).toHaveLength(0)
      })
    })
    test('should throw error if no customerID is passed', () => {
      expect(() =>
        customerErasure.getCustomerData()
      ).toThrowErrorMatchingSnapshot()
    })
  })

  describe('::deleteAll', () => {
    let payload
    beforeEach(() => {
      payload = {
        statusCode: 200,
        body: {
          results: [{ version: 1, id: 'id1' }, { version: 1, id: 'id2' }],
        },
      }
      customerErasure.client.execute = jest.fn(() => Promise.resolve(payload))
    })
    test('should delete data', async () => {
      const data = await customerErasure.getCustomerData('customerId')
      expect(data).toBeTruthy()
      await customerErasure.deleteAll('customerId')
    })
    test('should throw error if no customerID is passed', () => {
      expect(() => customerErasure.deleteAll()).toThrowErrorMatchingSnapshot()
    })
  })

  describe('::buildReference', () => {
    test('should build reference', () => {
      expect(
        CustomerErasure.buildReference(['id1', 'id2', 'id3'])
      ).toMatchSnapshot()
    })
  })

  describe('::buildRequest', () => {
    test('should build request', () => {
      expect(
        CustomerErasure.buildRequest('example.com', 'GET')
      ).toMatchSnapshot()
    })
  })
})
