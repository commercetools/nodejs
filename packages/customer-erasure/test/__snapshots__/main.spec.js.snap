// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerErasure ::buildReference should build reference 1`] = `"resource(id in (\\"id1\\", \\"id2\\", \\"id3\\"))"`;

exports[`CustomerErasure ::buildRequest should build request 1`] = `
Object {
  "method": "GET",
  "uri": "example.com",
}
`;

exports[`CustomerErasure ::constructor should throw error if no \`apiConfig\` in \`options\` parameter 1`] = `"The constructor must be passed an \`apiConfig\` object"`;

exports[`CustomerErasure ::deleteAll should throw error if no customerID is passed 1`] = `"missing \`customerId\` argument"`;

exports[`CustomerErasure ::getCustomerData should throw error if no customerID is passed 1`] = `"missing \`customerId\` argument"`;

exports[`CustomerErasure ::getCustomerData with status code 200 should fetch data 1`] = `
Array [
  Object {
    "id": "id1",
    "version": 1,
  },
  Object {
    "id": "id2",
    "version": 1,
  },
  Object {
    "id": "id1",
    "version": 1,
  },
  Object {
    "id": "id2",
    "version": 1,
  },
  Object {
    "id": "id1",
    "version": 1,
  },
  Object {
    "id": "id2",
    "version": 1,
  },
  Object {
    "id": "id1",
    "version": 1,
  },
  Object {
    "id": "id2",
    "version": 1,
  },
  Object {
    "id": "id1",
    "version": 1,
  },
  Object {
    "id": "id2",
    "version": 1,
  },
  Object {
    "id": "id1",
    "version": 1,
  },
  Object {
    "id": "id2",
    "version": 1,
  },
  Object {
    "id": "id1",
    "version": 1,
  },
  Object {
    "id": "id2",
    "version": 1,
  },
]
`;

exports[`CustomerErasure ::getCustomerData with status code 500 should throw internal server error 1`] = `"Request returned status code 500"`;
